cast/java/ecj/src/main/java/com/ibm/wala/cast/java/translator/jdt/JDTJava2CAstTranslator.java

```diff
@@ -929,6 +943,127 @@ else if ((n.getModifiers() & Modifier.ABSTRACT) != 0) {
     return new ProcedureEntity(mdast, n, classBinding, memberEntities, context, annotations);
   }
 
+  public CAstNode visit(LambdaExpression n, WalkContext oldContext) {
+
+    ITypeBinding typeBinding = n.resolveTypeBinding().getErasure();
+    List<?> parameters = n.parameters();
+    List<CAstType> castTypes = new ArrayList<>();
+    List<String> castNames = new ArrayList<>();
+
+    for (Object obj : parameters) {
+      VariableDeclarationFragment tmp = (VariableDeclarationFragment) obj;
+      CAstType td = fTypeDict.getCAstTypeFor(tmp.resolveBinding().getType());
+      castTypes.add(td);
+      String tmpName = tmp.getName().getIdentifier();
+      castNames.add(tmpName);
+    }
+
+    // pass in memberEntities to the context, later visit(New) etc. may add classes
+    final Map<CAstNode, CAstEntity> memberEntities = HashMapFactory.make();
+    final MethodContext context =
+        new MethodContext(oldContext, memberEntities); // LEFTOUT: in polyglot there is a
+    // class context in between method and
+    // root
+
+    CAstNode mdast = fFactory.makeNode(CAstNode.RETURN, visitNode(n.getBody(), context));
+
+    // Polyglot comment: Presumably the MethodContext's parent is a ClassContext,
+    // and he has the list of initializers. Hopefully the following
+    // will glue that stuff in the right place in any constructor body.
+
+    if (context.getNameDecls() != null && !context.getNameDecls().isEmpty()) {
+      // new first statement will be a block declaring all names.
+      mdast =
+          fFactory.makeNode(
+              CAstNode.BLOCK_STMT,
+              context.getNameDecls().size() == 1
+                  ? context.getNameDecls().iterator().next()
+                  : fFactory.makeNode(CAstNode.BLOCK_STMT, context.getNameDecls()),
+              mdast);
+    }
+
+    /* TODO:
+     * createClassDeclaration creates a Function type when instantiating
+     * the lambda subclass. This is not correct, so a new type will have
+     * to be implemented.
+     * */
```

Comment:
I don't understand this comment; is it still relevant?

Reply:
indeed, this is now wrong.  I'll remove it.,